### PR TITLE
Added `import` keyword

### PR DESCRIPTION
**Pull request by Wodann**
Added AST construction of the import keyword. Currently you can import:
- items
- paths
- trees
- aliases

Documentation and test cases have been updated accordingly.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -39,12 +39,22 @@ function internalFunction() {}
...
+import path::to::exportedFunction;
+```
+
 ## EBNF
 
 Grammar of the language so far. This is likely outdated and definitely not proper EBNF but it paints a picture.
 
 ```
-SourceFile ::= Declaration*
+SourceFile ::=
```
Comment (baszalmstra) [inline README.md:55]: Reading the implementation `ImportItem` should be a `DeclarationItem`.

I think this is currently also correct syntax.

```
export import ::foo; 
```

File: README.md [hunk 2]
```diff
@@ -88,6 +98,20 @@ Literal =
   | FLOAT_NUMBER
   | "true"
   | "false"
+ImportItem ::= ImportTree ";"
```
Comment (baszalmstra) [inline README.md:101]: This is missing the `"import"` keyword.

File: README.md [hunk 3]
```diff
@@ -88,6 +98,20 @@ Literal =
...
+  | Alias
+ImportTreeList ::= "{" ImportTreeListElements "}"
+ImportTreeListElements ::=
+    ImportTree
+  | ImportTree "," ImportTreeListElements
+Path ::=
+    PathSegment
+  | PathSegment "::" Path
+PathSegment ::= NameRef
+Alias ::= ImportTree "as" NameRef
```
Comment (baszalmstra) [inline README.md:114]: `Alias` is recursive on itself through `ImportTree` which makes `import ::foo as bar as foo` valid syntax.

File: README.md [hunk 4]
```diff
@@ -88,6 +98,20 @@ Literal =
...
+    ImportTreeList
+  | Path
+  | Alias
+ImportTreeList ::= "{" ImportTreeListElements "}"
+ImportTreeListElements ::=
+    ImportTree
+  | ImportTree "," ImportTreeListElements
+Path ::=
+    PathSegment
+  | PathSegment "::" Path
```
Comment (baszalmstra) [inline README.md:112]: This makes `import foo::bar::{a, foo::bar}` invalid syntax

File: crates/mun_syntax/src/parsing/grammar/paths.rs [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+fn path_segment(p: &mut Parser, first: bool) {
+    let m = p.start();
+    if first {
+        p.eat(COLONCOLON);
+    }
+    match p.current() {
+        IDENT => {
+            name_ref(p);
+        }
+        SELF_KW | SUPER_KW => p.bump(),
```
Comment (baszalmstra) [inline crates/mun_syntax/src/parsing/grammar/paths.rs:46]: This makes `super::super::self` a legal path.

**Review by baszalmstra — COMMENTED** [review]
This is my first pass over the pull requests. There are a bunch of cases that I think should be invalid syntax. Also the EBNF is incorrect and does not reflect the changes to the source 1 on 1. 

Maybe it would be useful to add tests for failed cases (cases that should not parse correctly) and/or "inline" tests like in Rust Analyzer.

File: crates/mun_syntax/src/parsing/grammar/paths.rs (hunk 5 above)
Comment (baszalmstra) [reply crates/mun_syntax/src/parsing/grammar/paths.rs:46]: This should probably be resolved on a higher level.

**Comment by baszalmstra** [comment]
I took the paths implementation from this PR

**Comment by Wodann** [comment]
Considering whether to update this PR or to close it. Did you take most of it already? @baszalmstra

**Comment by baszalmstra** [comment]
I took the path part, Im gonna merge the rest once I get around imports down the line. I would keep it open.

**Comment by Wodann** [comment]
Parts of this code have already been cannibalised and used in other commits, so I'll close the PR.

**Pull request closed without merging**